docs/esp32/pcnt.rst

```diff
@@ -0,0 +1,275 @@
+
+      - *phase_a*, *phase_b* are input pins :ref:`machine.Pin <machine.Pin>` for monitoring of quadrature encoder pulses.
+        They are required in the constructor.
+
+      - *x124* is a hardware multiplier, possible values is 1, 2, 4. The default value is 4.
+        More info in `Quadrature decoder state table <https://en.wikipedia.org/wiki/Incremental_encoder#Quadrature_decoder>`_.
+        When more Encoder resolution is needed, it is possible for the encoder to count the leading
+        and trailing edges of the quadrature encoder’s pulse train from one channel,
+        which doubles (x2) the number of pulses. Counting both leading and trailing edges
+        of both channels (A and B channels) of a quadrature encoder will quadruple (x4) the number of pulses:
+
+          - 1 - count the leading(or trailing) edges from one phase channel.
+          - 2 - count the leading and trailing edges from one phase channel.
+          - 4 - count both leading and trailing edges of both phase channels.
+
+      - *scale* sets the scale value. The default value is 1. You may treat the scale
+        factor as **click per impulse**, **revolution per impulse**, **angle per impulse** etc.
+        Hint: Set scale factor to 1/4 to balance the multiplier x124=4.
+
+    These keywords are the same as the Counter keywords, see above:
+      - *filter_ns*
+      - *match1*
+      - *match2*
+
+Methods
+-------
+
+.. method:: Encoder.init(*, phase_a, ...)
+
+   Modify the settings of the Encoder object. See the **Constructor** for details about the parameters.
+
+The Encoder has the same methods as the Counter and differs only
+in the constructor and internal hardware PCNT initialization.
+
+Constants
+---------
+
+.. data:: Encoder.IRQ_MATCH1
+          Encoder.IRQ_MATCH2
+          Encoder.IRQ_ZERO
+
+   Selects callback triggers.
+
+::
+
+    from machine import Encoder, Pin
+
+    try:
+        n = 0
+        def irq_handler1(self):
+            n -= 1
+            print('irq_handler1()', self.id(), self.value(), n)
+
+        def irq_handler2(self):
+            n += 1
+            print('irq_handler2()', self.id(), self.value(), n)
+
+        enc = Encoder(0, phase_a=Pin(17, mode=Pin.IN), phase_b=Pin(16, mode=Pin.IN), match1=-1000, match2=1000)
+
+        enc.pause()
+        flt = enc.filter()  # return current filter value.
+        enc.filter(10_000)  # filter delay is 10ms
+        c = enc.value(0)  # get current encoder value, set the encoder value to 0
+        cnt.irq(irq_handler1, Counter.IRQ_MATCH1)  # set irq handler
+        cnt.irq(irq_handler2, Counter.IRQ_MATCH2)  # set irq handler
+        enc.resume()
+
+        _c = None
+        while True:
+            c = enc.value()  # get the encoder value
+            if _c != c:
+                _c = c
+                print('Encoder =', c)
+    finally:
+        encoder.deinit()  # free the input pins and encoder.
```

Comment:
encoder is also undefined. Should probably be enc

Reply:
I've implemented the PCNT in a very simple manner at https://github.com/harbaum/micropython/blob/ESP32PCNT/ports/esp32/machine_pcnt.c

This adds less code to MP itself. The encoder demo from this PR can run with an Encoder.py which  can easily be subclassed this like:

`
# Encoder.py

from machine import PCNT, Pin

class Counter:
    IRQ_MATCH1 = PCNT.EVT_L_LIM
    IRQ_MATCH2 = PCNT.EVT_H_LIM

class Encoder(PCNT):
    def __init__(self, unit_id, phase_a, phase_b, match1, match2):
        super().__init__(unit_id)
        self.unit_id = unit_id

        # react on pin_a edges
        self.config(pin = phase_a, ctrl = phase_b, channel = 0, 
            pos_mode = PCNT.COUNT_DEC, neg_mode = PCNT.COUNT_INC, 
            lctrl_mode = PCNT.MODE_REVERSE, hctrl_mode = PCNT.MODE_KEEP,
            counter_l_lim = match1, counter_h_lim = match2)

        # react on pin_b edges
        self.config(pin = phase_b, ctrl = phase_a, channel = 1, 
            pos_mode = PCNT.COUNT_INC, neg_mode = PCNT.COUNT_DEC, 
            lctrl_mode = PCNT.MODE_REVERSE, hctrl_mode = PCNT.MODE_KEEP,
            counter_l_lim = match1, counter_h_lim = match2)

        self.counter = 0
        self.match_handler = { }
        super().irq(self.handler)

    def id(self):
        return self.unit_id

    def pause(self):
        self.counter_pause()

    def resume(self):
        self.counter_resume()

    def filter(self, *argv):
        # convert from nanoseconds into PCNT units
        if len(argv) > 0:
            self.filter_value(2*argv[0]//25)
        else:
            return 25*self.filter_value()//2

    def value(self, *argv):
        # set/get the counter value
        counter = self.counter
        count = self.counter_value()

        if len(argv) > 0:
            if argv[0]:
                self.counter = argv[0] - count
            else:
                self.counter_clear()
                self.counter = 0

        return counter + count

    def handler(self, obj):
        print("handler", obj)
        status = obj.event_status()
        for e in [ PCNT.EVT_L_LIM, PCNT.EVT_H_LIM ]:
            if status & e and e in self.match_handler:
                self.match_handler[e](self)

    def irq(self, handler, event):
        self.event_enable(event)
        self.match_handler[event] = handler
`